dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -256,136 +213,242 @@ protected void updateDBWorkflowWithSourceControlWorkflow(Workflow workflow, Work
+                throw new CustomWebApplicationException(msg, LAMBDA_FAILURE);
+            }
+        } catch (ClassCastException ex) {
+            String msg = "Could not parse .dockstore.yml";
+            LOG.info(msg, ex);
             throw new CustomWebApplicationException(msg, LAMBDA_FAILURE);
-        });
+        }
+        updatedServices.add(createWorkflowAndVersionFromDockstoreYml(Service.class, repository, gitReference, user, dockstoreYml, "", subclass, gitHubSourceCodeRepo));
+        return updatedServices;
+    }
 
-        // Get Installation Access Token
-        String installationAccessToken = GitHubHelper.gitHubAppSetup(gitHubAppId, gitHubPrivateKeyFile, installationId);
+    /**
+     * Create or retrieve workflow or service based on Dockstore.yml, also add or update tag version
+     * @param workflowType Either BioWorkflow.class or Service.class
+     * @param repository Repository path (ex. dockstore/dockstore-ui2)
+     * @param gitReference Tag reference from GitHub (ex. 1.0)
+     * @param user User that triggered action
+     * @param dockstoreYml Dockstore YAML File
+     * @param workflowName User that triggered action
+     * @param subclass Subclass of the workflow
+     * @param gitHubSourceCodeRepo Source Code Repo
+     * @return New or updated workflow
+     */
+    @SuppressWarnings({"checkstyle:ParameterNumber"})
+    private Workflow createWorkflowAndVersionFromDockstoreYml(Class workflowType, String repository, String gitReference, User user, SourceFile dockstoreYml, String workflowName, String subclass, GitHubSourceCodeRepo gitHubSourceCodeRepo) {
+        // Check for existing workflow
+        String dockstoreWorkflowPath = "github.com/" + repository + (workflowName != null && !workflowName.isEmpty() ? "/" + workflowName : "");
+        Optional<Workflow> workflow = workflowDAO.findByPath(dockstoreWorkflowPath, false, workflowType);
+
+        Workflow workflowToUpdate = null;
+        // Create workflow if one does not exist
+        if (workflow.isEmpty()) {
+            // Ensure that a Dockstore user exists to add to the workflow
+            if (user == null) {
+                String msg = "User does not have an account on Dockstore.";
+                LOG.info(msg);
+                throw new CustomWebApplicationException(msg, LAMBDA_FAILURE);
+            }
 
-        // Create a service object
-        final GitHubSourceCodeRepo sourceCodeRepo = (GitHubSourceCodeRepo)SourceCodeRepoFactory.createGitHubAppRepo(installationAccessToken);
+            if (workflowType == BioWorkflow.class) {
+                workflowToUpdate = gitHubSourceCodeRepo.initializeWorkflowFromGitHub(repository, subclass, workflowName);
+            } else if (workflowType == Service.class) {
+                workflowToUpdate = gitHubSourceCodeRepo.initializeServiceFromGitHub(repository, subclass);
+            } else {
+                LOG.error(workflowType.getCanonicalName()  + " is not a valid workflow type.");
```

Comment:
If you come in here, don't you want to thrown an exception? Otherwise it will go ahead and create the workflow on line 369.

Reply:
Good catch